src/auth/hooks.ts

```diff
@@ -0,0 +1,319 @@
+import {defineMessages, useIntl} from 'react-intl'
+import {Alert, AppState} from 'react-native'
+import {useMutation, UseMutationOptions, useQuery, useQueryClient, UseQueryOptions} from 'react-query'
+
+import {useMutationWithInvalidations, useWallet} from '../hooks'
+import globalMessages from '../i18n/global-messages'
+import {decryptData, encryptData} from '../legacy/commonUtils'
+import {WrongPassword} from '../legacy/errors'
+import {WalletMeta} from '../legacy/state'
+import storage from '../legacy/storage'
+import {SettingsStorageKeys, Storage} from '../Storage'
+import {WalletJSON, walletManager, YoroiWallet} from '../yoroi-wallets'
+import {Keychain} from './Keychain'
+import {AuthenticationPrompt, authOsEnabled} from './KeychainStorage'
+
+export const useAuthOsEnabled = (options?: UseQueryOptions<boolean, Error>) => {
+  const queryClient = useQueryClient()
+  const query = useQuery({
+    queryKey: ['authOsEnabled'],
+    queryFn: authOsEnabled,
+    suspense: true,
+    ...options,
+  })
+
+  React.useEffect(() => {
+    const appStateSubscription = AppState.addEventListener('change', async (appState) => {
+      // when using OS auth and app is active again needs to check if still enabled
+      if (appState === 'active') {
+        query.refetch()
+      }
+    })
+    return () => appStateSubscription?.remove()
+  }, [query, queryClient])
+
+  if (query.data == null) return false
+
+  return query.data
+}
+
+export const useEnableAuthWithOs = (
+  {authenticationPrompt, storage}: {authenticationPrompt?: AuthenticationPrompt; storage: Storage},
+  options?: UseMutationOptions<void, Error>,
+) => {
+  const {authWithOs: enableAuthWithOs, ...mutation} = useAuthWithOs(
+    {
+      ...options,
+      onSuccess: (data, variables, context) => {
+        storage
+          .setItem(SettingsStorageKeys.Auth, JSON.stringify(AUTH_WITH_OS))
+          .then(() => storage.getItem(SettingsStorageKeys.Pin))
+          .then((pin) => (pin != null ? storage.removeItem(SettingsStorageKeys.Pin) : undefined))
+        options?.onSuccess?.(data, variables, context)
+      },
+    },
+    {authenticationPrompt},
+  )
+
+  return {...mutation, enableAuthWithOs}
+}
+
+export const useAuthWithOs = (
+  options?: UseMutationOptions<void, Error>,
+  {authenticationPrompt}: {authenticationPrompt?: AuthenticationPrompt} = {},
+) => {
+  const strings = useStrings()
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const alert = (error: unknown) => {
+    if (error instanceof Keychain.Errors.CancelledByUser) return
+    if (error instanceof Keychain.Errors.TooManyAttempts) return Alert.alert(strings.error, strings.tooManyAttempts)
+    return Alert.alert(strings.error, strings.unknownError)
+  }
+
+  const defaultAuthenticationPrompt: AuthenticationPrompt = {
+    cancel: strings.cancel,
+    title: strings.authorize,
+  }
+
+  const mutation = useMutationWithInvalidations({
+    ...options,
+    invalidateQueries: [['authSetting']],
```

Comment:
why does this need to invalidate?

Reply:
part as success process

Reply:
https://github.com/Emurgo/yoroi-mobile/pull/2213/commits/e77da4fc7c883dad571b864fc5eb8d68278af5a4